### PR TITLE
refactor(compiler): introduce directive wrappers to generate less code

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+const CURR_VALUE_VAR = o.variable('currValue');
+const THROW_ON_CHANGE_VAR = o.variable('throwOnChange');
+const FORCE_UPDATE_VAR = o.variable('forceUpdate');
+const VIEW_VAR = o.variable('view');
+const RENDER_EL_VAR = o.variable('el');
+
+const RESET_CHANGES_STMT = o.THIS_EXPR.prop(CHANGES_FIELD_NAME).set(o.literalMap([])).toStmt();
+
+@Injectable()
+export class DirectiveWrapperCompiler {
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: inline comments

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 2]
```diff
@@ -69,8 +69,8 @@ export class OfflineCompiler {
   }
 
   compile(
-      moduleUrl: string, ngModulesSummary: NgModulesSummary, components: StaticSymbol[],
-      ngModules: StaticSymbol[]): Promise<SourceModule[]> {
+      moduleUrl: string, ngModulesSummary: NgModulesSummary, directives: StaticSymbol[],
+      components: StaticSymbol[], ngModules: StaticSymbol[]): Promise<SourceModule[]> {
```
Comment (vicb) [inline modules/@angular/compiler/src/offline_compiler.ts]: merge dir & cmp ?

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 3]
```diff
@@ -80,11 +80,15 @@ export class OfflineCompiler {
     exportedVars.push(
         ...ngModules.map((ngModuleType) => this._compileModule(ngModuleType, statements)));
 
+    // compile directive wrappers
+    exportedVars.push(...directives.map(
+        (directiveType) => { return this._compileDirectiveWrapper(directiveType, statements); }));
```
Comment (vicb) [inline modules/@angular/compiler/src/offline_compiler.ts]: nit: `{ return`

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 4]
```diff
@@ -230,81 +248,46 @@ export function bindDirectiveHostProps(
 
 export function bindDirectiveInputs(
     directiveAst: DirectiveAst, directiveInstance: o.Expression, compileElement: CompileElement) {
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: rename `directiveInstance` to `directiveWrapperInstance`

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    const ctorStmts: o.Statement[] =
+        [o.THIS_EXPR.prop(CHANGED_FIELD_NAME).set(o.literal(false)).toStmt()];
+    if (lifecycleHooks.genChanges) {
+      ctorStmts.push(RESET_CHANGES_STMT);
+    }
+
+    const fields: o.ClassField[] = [
+      new o.ClassField(CONTEXT_FIELD_NAME, o.importType(dirMeta.type)),
+    ];
+    if (lifecycleHooks.genChanges) {
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Merge with previous `if (genChanges)`

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+      fields.push(new o.ClassField(CHANGES_FIELD_NAME, new o.MapType(o.DYNAMIC_TYPE)));
+    }
+    const methods: o.ClassMethod[] = [];
+    const inputFieldNames: string[] = Object.keys(dirMeta.inputs);
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Inline

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    o.THIS_EXPR.prop(CONTEXT_FIELD_NAME).prop(input).set(CURR_VALUE_VAR).toStmt(),
+  ];
+  if (lifecycleHooks.genChanges) {
+    onChangeStatements.push(o.THIS_EXPR.prop(CHANGES_FIELD_NAME)
+                                .key(o.literal(input))
+                                .set(o.importExpr(resolveIdentifier(Identifiers.SimpleChange))
+                                         .instantiate([fieldExpr, CURR_VALUE_VAR]))
+                                .toStmt());
+  }
+  onChangeStatements.push(fieldExpr.set(CURR_VALUE_VAR).toStmt());
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Move up

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+  const changedVar = o.variable('changed');
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Remove variable.

**Comment by tbosch** [comment]
Verify performance impact on tree and table (create, update, destroy), should not have any.

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

File: modules/@angular/compiler/src/identifiers.ts [hunk 9]
```diff
@@ -259,12 +262,22 @@ export class Identifiers {
...
   };
   static TRANSLATIONS_FORMAT: IdentifierSpec = {
     name: 'TRANSLATIONS_FORMAT',
     moduleUrl: assetUrl('core', 'i18n/tokens'),
-    runtime: TRANSLATIONS_FORMAT_
+    runtime: TRANSLATIONS_FORMAT
+  };
+  static setBindingDebugInfo: IdentifierSpec = {
+    name: 'setBindingDebugInfo',
+    moduleUrl: assetUrl('core', 'i18n/tokens'),
```
Comment (tbosch) [inline modules/@angular/compiler/src/identifiers.ts]: Done

File: modules/@angular/compiler/src/offline_compiler.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Done

File: modules/@angular/compiler/src/offline_compiler.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Done

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 10]
```diff
@@ -31,15 +31,18 @@ function createCurrValueExpr(exprIndex: number): o.ReadVarExpr {
...
-function bind(
-    view: CompileView, currValExpr: o.ReadVarExpr, fieldExpr: o.ReadPropExpr,
-    parsedExpression: cdAst.AST, context: o.Expression, actions: o.Statement[],
-    method: CompileMethod, bindingIndex: number) {
+class EvalResult {
+  constructor(public forceUpdate: o.Expression) {}
+}
+
+function evalCdAst(
+    view: CompileView, currValExpr: o.ReadVarExpr, parsedExpression: cdAst.AST,
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: I keep it for now

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Can't as we need to keep the old value

File: modules/@angular/compiler/src/directive_wrapper_compiler.ts (hunk 8 above)
Comment (tbosch) [inline modules/@angular/compiler/src/directive_wrapper_compiler.ts]: Done

**Comment by tbosch** [comment]
Checked the numbers on the tree benchmark, no regression for `create` nor `update`

**Comment by tbosch** [comment]
Reviewed in person with @vicb

**Comment by tbosch** [comment]
Note: The failure is due to our browserstack license.

**Comment by alxhub** [comment]
This is failing in g3sync, likely due to build changes that are needed on the g3 side. I'll work with @alexeagle on getting this in.

**Comment by alxhub** [comment]
Merged as b0a03fc.

**Pull request closed without merging**